src/engine/training_test.ts

```diff
@@ -889,6 +889,67 @@ describeMathCPUAndGPU('Model.fit', () => {
         layer2.getWeights()[0], mul(scalar(-0.11295), ones([10, 1])));
   });
 
+  it('Setting trainable of layer from fit callback', async () => {
+    const model = tfl.sequential();
+    model.add(
+        tfl.layers.dense({units: 3, activation: 'relu', inputShape: [4]}));
+    model.add(tfl.layers.dense({units: 1}));
+    model.compile({loss: 'meanSquaredError', optimizer: 'sgd'});
+
+    expect(model.trainableWeights.length).toEqual(4);
+    const xs = tfc.randomNormal([5, 4]);
+    const ys = tfc.randomNormal([5, 1]);
+    const layer1KernelValues: Float32Array[] = [];
+    const layer2KernelValues: Float32Array[] = [];
+    await model.fit(xs, ys, {
+      epochs: 3,
+      callbacks: {
+        onEpochEnd: async (epoch, logs) => {
+          layer1KernelValues.push(
+              model.layers[0].getWeights()[0].dataSync() as Float32Array);
+          layer2KernelValues.push(
+              model.layers[1].getWeights()[0].dataSync() as Float32Array);
+          // Freeze the first dense layer after the 2nd epoch and unfreeze it
+          // after the 3rd epoch.
+          if (epoch === 1) {
+            model.layers[0].trainable = false;
+          } else if (epoch === 2) {
+            model.layers[0].trainable = true;
+          }
+          if (epoch > 0) {
+            // The 2nd dense layer is never frozen. So its kernel should
```

Comment:
Specifically, this checks that the largest weight value, by absolute value, has changed, and is a proxy for a predicate testing that the weights are different, right?   Would `!expectArraysClose` work better here and below?

Reply:
This doesn't check the change in the value with the largest absolute value. Instead, it checks the largest change in any value, by absolute value. So if any value changed, it'll be reflected in this.

Unfortunately, there is no `expectArraysNotClose` in the testing library right now. The approach here is the workaround.

Reply:
Yeah, sorry, inaccurate langauge.  "the largest change in value is greater than zero".

Is  expect(expectArraysClose(a, b)).toThrow(text));

better?

Should we add expectArraysDifferInValue to core?

Reply:
drive by review: I feel like expectArraysDiffer is a slight test code smell. better to fix a seed and expect exact equality to avoid a flaky test?